File: wpimath/src/main/java/org/wpilib/math/trajectory/struct/TrajectorySampleStruct.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.trajectory.struct;

import java.nio.ByteBuffer;
import org.wpilib.math.trajectory.TrajectorySample;
import org.wpilib.util.struct.Struct;

public class TrajectorySampleStruct implements Struct<TrajectorySample> {
  @Override
  public Class<TrajectorySample> getTypeClass() {
    return TrajectorySample.class;
  }

  @Override
  public String getTypeName() {
    return "TrajectorySample.Base";
  }

  @Override
  public int getSize() {
    return DOUBLE_SIZE;
  }

  @Override
  public String getSchema() {
    return "double time";
  }

  @Override
  public TrajectorySample unpack(ByteBuffer bb) {
    double time = bb.getDouble();
    return new TrajectorySample(time);
  }

  @Override
  public void pack(ByteBuffer bb, TrajectorySample value) {
    bb.putDouble(value.time);
  }
}

File: wpimath/src/main/native/cpp/trajectory/struct/TrajectorySampleStruct.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/trajectory/struct/TrajectorySampleStruct.hpp"

namespace {
constexpr size_t TIMESTAMP_OFF = 0;
}  // namespace

using StructType = wpi::util::Struct<wpi::math::TrajectorySample>;

wpi::math::TrajectorySample StructType::Unpack(std::span<const uint8_t> data) {
  return wpi::math::TrajectorySample{wpi::units::seconds<>{
      wpi::util::UnpackStruct<double, TIMESTAMP_OFF>(data)}};
}

void StructType::Pack(std::span<uint8_t> data,
                      const wpi::math::TrajectorySample& value) {
  wpi::util::PackStruct<TIMESTAMP_OFF>(data, value.time.value());
}
